internal/app/orchestrator/orchestrator.go

```diff
@@ -90,16 +105,172 @@ func New(ctx context.Context, l log.Logger, mapper mapper.Mapper, upstreamClient
 //
 // Cancel is an optional function to release resources in the producer. If
 // provided, the consumer may call this function multiple times.
-func (c *orchestrator) CreateWatch(req gcp.Request) (chan gcp.Response, func()) {
-	// TODO implement.
-	return nil, nil
+func (o *orchestrator) CreateWatch(req gcp.Request) (chan gcp.Response, func()) {
+	ctx := context.Background()
```

Comment:
The interface of gcp does not have the context as a parameter. https://github.com/envoyproxy/go-control-plane/blob/f070f3d1d73b03f00369ee424d38077944e7b984/pkg/cache/v2/cache.go#L43

Reply:
Yeah ctx can be passed from constructor. We will need that to make sure clean up happens on shutdown

Reply:
How would you pass the ctx from the constructor? The constructor doesn't call CreateWatch.

Apologies if I am misinterpreting you. But the only way I can see getting the context from the constructor is to store the context in the orchestrator object, but that is bad practice. [The context should be in the scope of the request, not the creation of the constructor.](https://dave.cheney.net/2017/08/20/context-isnt-for-cancellation) In an ideal world, CreateWatch should pass a context as a parameter. 

I agree that a cancel function needs to be set up and have each of the watch listen for it. Unclear what the best mechanism is as go-control-plane doesn't pass around ctx as a parameter in their interfaces.

Reply:
Thanks for the link. I will explore the possibility of using the `tomb` package. However,fortunately, because the xdsclient is initialized in [server.go](https://github.com/envoyproxy/xds-relay/blob/master/internal/app/server/server.go#L36), i think we are covered here. Nonetheless the same issue will be a problem in xds client.

Reply:
Looks like g-c-p [saves the context](https://github.com/envoyproxy/go-control-plane/blob/master/pkg/server/v2/server.go#L87).

Reply:
Thinking about this more, the [OpenStream](https://github.com/envoyproxy/xds-relay/blob/master/internal/app/upstream/client.go#L28) takes a context, which needs to be wired up to global context, so that cancellation caused the waiting goroutines for [RecvMsg](https://github.com/grpc/grpc-go/blob/b02de00073bfccaef63816bfd3ac296e9abc371f/stream.go#L126-L132) to exit.

Reply:
This isn't ideal. If you look at the [context package](https://golang.org/pkg/context/), they also state `Do not store Contexts inside a struct type; instead, pass a Context explicitly to each function that needs it. The Context should be the first parameter, typically named ctx`. 

I think it is worth refactoring go-control-plane such that we don't store it on the server but either pass it in or store a `Cancel` method in the server which we can call. But that is up to you.

Reply:
Yeah I agree that context shouldn't be saved. I'm suggesting that we need some way to cleanup resources eg connections ,goroutines etc during shutdown and other scenarios.

Reply:
I've been thinking more about this and have a [TODO](https://github.com/envoyproxy/xds-relay/pull/53/files#diff-a3a788e64cea038d60dbf226fa6e275eR37) in the upstream map clean up portion that relates to this.

An approach I've been considering is having `OpenStream` from the upstream client supply not only the receiver channel, but a "done" channel. 

```
OpenStream(context.Context, *v2.DiscoveryRequest, string) <-chan *Response, chan bool
```

when shutdown occurs, orchestrator will close the bool channel. Orchestrator [already has to do this](https://github.com/envoyproxy/xds-relay/pull/53/files#diff-a3a788e64cea038d60dbf226fa6e275eR30) to clean up the go routines during [watchUpstream](https://github.com/envoyproxy/xds-relay/pull/53/files#diff-2958638164db92742469d2bb08d0001aR232). Rather than initializing a new channel, we can just assign it the one returned from `OpenStream`.

Reply:
I could be wrong, but one aspect to consider while using channels for cancellation is that channels can have only one receiver. If there are more goroutines than one, we have to think about how to cancel all of them using the same channel. A context on the other hand can be used to send cancellation signal to multiple goroutines.

Reply:
Made a simple test here that shows closing the channel can notify multiple receivers: https://play.golang.org/p/QHhcOZXNLSa

Reply:
Thanks for the experiment. It could work out then.
I was using https://play.golang.org/p/NQGO5-jCVz to make my observations. Observe how only one goroutine gets all the values.

Reply:
Yeah, `close()` is a special case that seems to work differently than values being pushed to the channel (which only the faster receiver can read from)

Reply:
Reading about context cancellation, there's so many contradicting views. [here](https://dave.cheney.net/2017/08/20/context-isnt-for-cancellation), [here](https://twitter.com/davecheney/status/894717817886433280), [here](https://faiface.github.io/post/context-should-go-away-go2/).